### ISSUE TITLE
A getter and setter with different types for a property are not handled

### ISSUE THREAD

**Issue by viblo** [microsoft/pyright#993]
**Describe the bug**

Sometimes its useful to have a property where the getter and setter have different types. Seems like pyright does not handle this?

**Expected behavior**
No errors in the below code.

**Screenshots or Code**

```python
class B:
    @property
    def f(self) -> int:
        return 1
    @f.setter 
    def f(self, x: str) -> None:
        pass
b = B()
b.f = ''  
b.f + 1 # Error: Operator "+" not supported for types "str" and "Literal[1]"Pylance (reportGeneralTypeIssues)
```

**VS Code extension or command-line**
Running VS Code with pyright included with Pylance.
Pyright version: 1.1.65

**Comment by erictraut** [comment]
Using different types for a setter and deleter is a very bad practice. It is generally disallowed in all typed languages for good reason. Properties are meant to act like member variables, which do not have different types depending on whether you read or write them.

While this practice is "supported" in Python, it's only because Python historically didn't support static type declarations, so it had no way to enforce consistency here.

I had actually planned to add a diagnostic rule in Pyright that checks for mismatched types in getters/setters/deleters and flag them as an error.

Using different types in a getter and a setter not only makes code harder to understand, it also breaks some very important and useful assumptions that a static type checker makes. When a value is written to an object member, static type checkers (including Pyright) generally assume that the same type will be returned if the object member is read a second time. This allows for type narrowing on members.

Here is an example of such type narrowing:
```python
class MyClass:
    @property
    def var1(self) -> Union[int, str]: ...
    @var1.setter
    def var1(self, value: Union[int, str]): ...

a = MyClass()

a.var1 = 3
b = a.var1
reveal_type(b) # int

a.var1 = "hi"
c = a.var1
reveal_type(c) # str
```

I'll need to think about this more, but my current thought is to do both of the following:
1. Add a diagnostic rule that checks for mismatched types between property getter/setter/deleter. Enable this diagnostic rule by default in "strict" mode but leave it off by default in "basic" mode.
2. Add logic to detect getter/setter/deleter type mismatches and disable member type narrowing only for those members.

**Comment by viblo** [comment]
In my case I have a physics library with a vector class that I try to add type hints to, where many properties can be set by tuples or vector, but always return the vector.

For example body class:

```python
class Body:
    _position: Vec2d

    @property
    def position(self) -> Vec2d:
        return self._position

    @position.setter
    def position(self, v: Union[Tuple[float, float], Vec2d]) -> None:
        if isinstance(v, Vec):
            self._position = v
        else:
            self._position = Vec(v[0], v[1])
```

used in this way

```python
b = Body()
b.position = Vec2d(1, 2)
b.position += (3, 4) 
print(b.position.x)
```

Not sure if you follow the development of / look at what mypy does, but the same issue is reported there, also with some discussion: https://github.com/python/mypy/issues/3004

**Comment by erictraut** [comment]
Thanks for pointing out the thread in mypy. I agree with Guido that using different types for getters/setters/deleters is a perversion (to use his word). I would call it an "egregious abuse of a language feature".

After thinking about this a bit more, I think this should be flagged by a type checker as a type consistency violation. If you are choosing to use type annotations within your code, then it's generally the case that you care about using language features in the way they were intended.

So here's my updated plan:
1. Add a new diagnostic rule that checks for mismatched types between property getter/setter/deleter. Enable this diagnostic rule by default in both "basic" and "strict" mode.
2. Do nothing to handle narrowing in a special way. If you want to abuse properties in this manner, you can use an `Any` type or use `# type: ignore` comments to suppress the resulting errors.

**Comment by erictraut** [comment]
This is implemented in Pyright 1.1.67, which I just published. It will also be included in next week's release of Pylance.
